src/thread.rs

```diff
@@ -0,0 +1,483 @@
+            "Bug: multiple threads try to join the same thread."
+        );
+        if self.threads[joined_thread_id].state != ThreadState::Terminated {
+            // The joined thread is still running, we need to wait for it.
+            self.active_thread_mut().state = ThreadState::BlockedOnJoin(joined_thread_id);
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Allocate a new blockset id.
+    fn create_blockset(&mut self) -> BlockSetId {
+        self.blockset_counter = self.blockset_counter.checked_add(1).unwrap();
+        self.blockset_counter.into()
+    }
+
+    /// Block the currently active thread and put it into the given blockset.
+    fn block_active_thread(&mut self, set: BlockSetId) {
+        let state = &mut self.active_thread_mut().state;
+        assert_eq!(*state, ThreadState::Enabled);
+        *state = ThreadState::Blocked(set);
+    }
+
+    /// Unblock any one thread from the given blockset if it contains at least
+    /// one. Return the id of the unblocked thread.
+    fn unblock_random_thread(&mut self, set: BlockSetId) -> Option<ThreadId> {
+        for (id, thread) in self.threads.iter_enumerated_mut() {
+            if thread.state == ThreadState::Blocked(set) {
+                trace!("unblocking {:?} in blockset {:?}", id, set);
+                thread.state = ThreadState::Enabled;
+                return Some(id);
+            }
+        }
+        None
+    }
+
+    /// Change the active thread to some enabled thread.
+    fn yield_active_thread(&mut self) {
+        self.yield_active_thread = true;
+    }
+
+    /// Decide which action to take next and on which thread.
+    fn schedule(&mut self) -> InterpResult<'tcx, SchedulingAction> {
+        if self.threads[self.active_thread].check_terminated() {
+            // Check if we need to unblock any threads.
+            for (i, thread) in self.threads.iter_enumerated_mut() {
+                if thread.state == ThreadState::BlockedOnJoin(self.active_thread) {
+                    trace!("unblocking {:?} because {:?} terminated", i, self.active_thread);
+                    thread.state = ThreadState::Enabled;
+                }
+            }
+            return Ok(SchedulingAction::ExecuteDtors);
+        }
+        if self.threads[self.active_thread].state == ThreadState::Enabled
+            && !self.yield_active_thread
+        {
+            return Ok(SchedulingAction::ExecuteStep);
+        }
+        for (id, thread) in self.threads.iter_enumerated() {
+            if thread.state == ThreadState::Enabled {
+                if !(self.yield_active_thread && id == self.active_thread) {
```

Comment:
```suggestion
                if !self.yield_active_thread || id != self.active_thread {
```
I find this clearer.